data/src/main/java/org/apache/iceberg/data/ContentFileWriteBuilderImpl.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.MetricsConfig;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SortOrder;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.encryption.EncryptionKeyMetadata;
+import org.apache.iceberg.io.DataWriter;
+import org.apache.iceberg.io.DeleteSchemaUtil;
+import org.apache.iceberg.io.WriteBuilder;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.ArrayUtil;
+
+/**
+ * An internal implementation that handles all {@link ContentFileWriteBuilder} interface variants.
+ *
+ * <p>This unified implementation serves as a backend for multiple specialized content writers:
+ *
+ * <ul>
+ *   <li>{@link DataWriteBuilder} for creating data files
+ *   <li>{@link EqualityDeleteWriteBuilder} for creating equality delete files
+ *   <li>{@link PositionDeleteWriteBuilder} for creating position delete files
+ * </ul>
+ *
+ * <p>The implementation delegates to a format-specific {@link WriteBuilder} while enriching it with
+ * content-specific functionality. When building a writer, the implementation configures the
+ * underlying builder and calls its {@link WriteBuilder#build()} method to create the appropriate
+ * specialized writer for the requested content type.
+ *
+ * @param <C> the concrete builder type for method chaining
+ * @param <W> the type of the wrapped format-specific writer builder
+ * @param <E> the engine-specific schema type required by the writer
+ */
+@SuppressWarnings("unchecked")
+class ContentFileWriteBuilderImpl<
+        C extends ContentFileWriteBuilderImpl<C, W, E>, W extends WriteBuilder<W, E>, E>
+    implements DataWriteBuilder<C, E>,
```

Comment:
does it make sense to have those 3 additional interfaces? can they be merged into the `ContentFileWriteBuilder` interface if these 3 builder interfaces aren't needed outside the data module?